### DISCUSSION TITLE
Misleading Error Message

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#13414] [Q&A] [answered]
Want some confirmation before i convert this discussion into an issue.

It seems that certain characters aren't allowed within string, and will cause unexpected error upon compilation.

```axaml
<Button Content="<-" Command="" Grid.Column="0" />
```

This should be valid, but it is apparently not as Avalonia will throw an error during the parsing process : 
```log
Touch-Gestures\Touch-Gestures.UX\Views\MainView.axaml : Avalonia error AVLN:0001: File doesn't contain valid XAML: System.Xml.XmlException:
 '<', hexadecimal value 0x3C, is an invalid attribute character. Line 19, position 25.
at System.Xml.XmlTextReaderImpl.Throw(Exception e)
at System.Xml.XmlTextReaderImpl.Throw(String res, String[] args)
at System.Xml.XmlTextReaderImpl.ParseAttributeValueSlow(Int32 curPos, Char quoteChar, NodeData attr)
at System.Xml.XmlTextReaderImpl.ParseAttributes()
at System.Xml.XmlTextReaderImpl.ParseElement()
at System.Xml.XmlTextReaderImpl.ParseElementContent()
at System.Xml.Linq.XContainer.ReadContentFrom(XmlReade r r)
at System.Xml.Linq.XDocument.Load(XmlReader reader, LoadOptions options)
at System.Xml.Linq.XDocument.Parse(String text, LoadOptions options)
at Avalonia.Build.Tasks.XamlFileInfo.Parse(String data) in /_/src/Avalonia.Build.Tasks/XamlFileInfo.cs:line 11 
at Avalonia.Build.Tasks.GenerateAvaloniaResourcesTask.PreProcessXamlFiles(List`1 sources) in /_/src/Avalonia.Build.Tasks/GenerateAvaloniaResourcesTask.cs:line 96 
```

I'm pretty sure that `<` & `>` aren't the only characters that would cause such issues, escaping characters seems also not possible within axaml.

**Comment by Sorien** [comment, marked as answer]
https://www.w3.org/TR/xml/#charsets The ampersand character (&) and the left angle bracket (<) must not appear in their literal form, except when used as markup delimiters, or within a [comment](https://www.w3.org/TR/xml/#dt-comment), a [processing instruction](https://www.w3.org/TR/xml/#dt-pi), or a [CDATA section](https://www.w3.org/TR/xml/#dt-cdsection)

**Reply by Mrcubix** [reply]
Then a new error message is in order to properly indicate that matter

**Reply by robloo** [reply]
@Mrcubix 

> Touch-Gestures\Touch-Gestures.UX\Views\MainView.axaml : Avalonia error AVLN:0001: File doesn't contain valid XAML: System.Xml.XmlException: '<', hexadecimal value 0x3C, is an invalid attribute character. Line 19, position 25.

Tells you
 1. The file
 2. The issue: Invalid XAML due to '<' usage (with hex)
 3. Where in the file the issue occurs

This is a high quality error message and there is no need to improve it.

**Reply by Mrcubix** [reply]
it says that as an attribute character, not as part of the content of a value

**Reply by robloo** [reply]
"attribute" vs. "attribute value" is almost semantically irrelevant. XML has had this requirement since inception and I think the compiler is pretty clear about it.

**Reply by Mrcubix** [reply]
you just said the keyword, `I` and peoples aren't you, if i had to ask the question, other peoples will wonder the same, it's about making the experience better for everyone, not those that worked with xml for a long time, it's about making it user-friendly, need i say more?

**Reply by timunie** [reply]
@Mrcubix what error message would you think is better? Any proposal?

**Reply by robloo** [reply]
Note that the exception is very likely rethrown from System.xml. It is the same exception Microsoft uses in .NET itself. Which also means we can't change that portion of it.

**Reply by Mrcubix** [reply]
> @Mrcubix what error message would you think is better? Any proposal?

If it is possible, and it should be, with some changes that is, apparently : 

```
The character '<' cannot be present within an attribute's value, The character may need to be escaped.
See [Link to a documentation page containing comprehensive list of characters] for more information.
```

I do not think that the link that robloo provided is appropriate for proper visualisation of those characters.

After doing some research (which may not be sufficient), i was not able to find a page that meet such criteria.
Most only provide character ranges, though one may convert those missing ranges into a table of illegal, and 'may be escaped' characters respectively.
